test/cpp/api/modulelist.cpp

```diff
@@ -0,0 +1,295 @@
+#include <gtest/gtest.h>
+
+#include <torch/nn/modules.h>
+#include <torch/nn/modules/batchnorm.h>
+#include <torch/nn/modules/conv.h>
+#include <torch/nn/modules/dropout.h>
+#include <torch/nn/modules/linear.h>
+#include <torch/nn/modules/modulelist.h>
+#include <torch/nn/modules/rnn.h>
+#include <torch/types.h>
+#include <torch/utils.h>
+
+#include <algorithm>
+#include <memory>
+#include <vector>
+
+#include <test/cpp/api/support.h>
+
+using namespace torch::nn;
+using namespace torch::test;
+
+struct ModuleListTest : torch::test::SeedingFixture {};
+
+TEST_F(ModuleListTest, ConstructsFromSharedPointer) {
+  struct M : torch::nn::Module {
+    explicit M(int value_) : value(value_) {}
+    int value;
+  };
+  ModuleList list(
+      std::make_shared<M>(1), std::make_shared<M>(2), std::make_shared<M>(3));
+  ASSERT_EQ(list->size(), 3);
+}
+
+TEST_F(ModuleListTest, ConstructsFromConcreteType) {
+  static int copy_count;
+
+  struct M : torch::nn::Module {
+    explicit M(int value_) : value(value_) {}
+    M(const M& other) : torch::nn::Module(other) {
+      copy_count++;
+    }
+    int value;
+  };
+
+  copy_count = 0;
+  ModuleList list(M(1), M(2), M(3));
+  ASSERT_EQ(list->size(), 3);
+  // NOTE: The current implementation expects each module to be copied exactly
+  // once, which happens when the module is passed into `std::make_shared<T>()`.
+  // TODO: Find a way to avoid copying, and then delete the copy constructor of
+  // `M`.
+  ASSERT_EQ(copy_count, 3);
+}
+
+TEST_F(ModuleListTest, ConstructsFromModuleHolder) {
+  struct MImpl : torch::nn::Module {
+    explicit MImpl(int value_) : value(value_) {}
+    int value;
+  };
+
+  struct M : torch::nn::ModuleHolder<MImpl> {
+    using torch::nn::ModuleHolder<MImpl>::ModuleHolder;
+    using torch::nn::ModuleHolder<MImpl>::get;
+  };
+
+  ModuleList list(M(1), M(2), M(3));
+  ASSERT_EQ(list->size(), 3);
+}
+
+TEST_F(ModuleListTest, PushBackAddsAnElement) {
```

Comment:
We should likely also add similar tests for `insert`, in order to test its support for different module types.